Decorator to show structural load representation in 3D view #3760

LM-Nascimento opened the pull request:
https://github.com/user-attachments/assets/5d794bea-ccaa-4545-aa5d-38e27427eb48

This can show the structural loads in 3D view, still has a few issues but it kinda works.

LM-Nascimento committed:
Initial commit, able to show structural curve action on the sample file of IFC documentation

LM-Nascimento committed:
updates to shader class

LM-Nascimento committed:
update decorator to match shader + changes to text rendering

LM-Nascimento committed:
now with shader for point loads

LM-Nascimento committed:
Merge branch 'IfcOpenShell:v0.8.0' into v0.8.0

LM-Nascimento committed:
shit ton of changes

LM-Nascimento committed:
Merge branch 'IfcOpenShell:v0.8.0' into v0.8.0

LM-Nascimento committed:
seems to be working

LM-Nascimento committed:
Merge branch 'v0.8.0' of https://github.com/LM-Nascimento/bonsaiDecoratorForLoads into v0.8.0

LM-Nascimento committed:
Cleaning up and type hints

LM-Nascimento committed:
separate shader file from decoration data

LM-Nascimento committed:
Merge branch 'IfcOpenShell:v0.8.0' into v0.8.0

LM-Nascimento committed:
cleanup

LM-Nascimento committed:
Merge branch 'v0.8.0' of https://github.com/LM-Nascimento/bonsaiDecoratorForLoads into v0.8.0

LM-Nascimento committed:
licence comment

LM-Nascimento committed:
Update unit.py

LM-Nascimento committed:
black formatting and changes in shader typing

LM-Nascimento committed:
Merge branch 'v0.8.0' of https://github.com/LM-Nascimento/bonsaiDecoratorForLoads into v0.8.0

LM-Nascimento committed:
black formatting

theoryshaw commented:
very cool!
would be awesome to incorporate something like this into BB, as well. 
https://community.osarch.org/discussion/2628/anastruct-2d-structural-analysis-in-python/p1

theoryshaw commented:
can you share your `structural-curve-member - with_point.ifc` file?  Would like to play around with it, if possible.

LM-Nascimento commented:
[structural-curve-member - with_point.ifc.txt](https://github.com/user-attachments/files/18189547/structural-curve-member.-.with_point.ifc.txt)
This is the file that I used. It is the example file in the IFC documentation, I added a point load and modified values for the nominal load (it is possible to edit the values using blender bim UI).

I also tested it with building 01 and building 02 from https://github.com/IfcOpenShell/analysis-models/tree/master/ifcFiles

I think we should improve structural model authoring workflow before attempting to integrate structural analysis into BB.

Moult commented:
Incredible! It's also an absolute behemoth of a PR so it's difficult to review, so I'm opting for the James Bond "merge first ask questions later" approach.

At a glance, I suspect there is a possibility to do some work in util.unit and use that more - right now it looks as though there is some duplication going on.

The `load_decoration_data.py` also sticks out as an anomaly - it's a new file that no other module has a similar situation. It doesn't mean it's wrong, just that it's something new :) Maybe there is also the ability have shared decoration code (if at all relevant).

Agree that structural model authoring needs to improve first before dealing with structural analysis.

LM-Nascimento commented:
Sorry about the size, I wasn't sure I could make it work, so I wanted to have a working prototype, that turned out to be huge. I have no background in programming, so I suppose that there is a more concise way to make this.
There is indeed duplication in the units part, the duplicate code in this PR isn't used, I just forgot to delete it.
load_decoration_data.py is responsible for getting the necessary information from the loads and members, there are many different types of loads, especially in curve members, so I decided to make a separate file due to the sheer number of lines. This decoration data  is very dependant on the custom shaders I wrote, so it would be difficult to share with other decoration code.
